Fix Issues With SWG Nuclear Missile/Cruise Missile/ICBM

commy2 opened the pull request:
### 1.04

- missile flies backwards/sideways

https://github.com/TheSuperHackers/GeneralsGamePatch/assets/6576312/a2448b76-85c2-4505-9fd8-47ce81e3d6ec



### this patch

- missile flies forwards

https://github.com/TheSuperHackers/GeneralsGamePatch/assets/6576312/4abd0743-f38e-430c-9b0c-b980dd159fff

changes:
  - fix: Cruise Missile is now fixed.
  - fix: ICBM is now fixed.

subchanges:
  - fix: The ICBM launcher now uses a unique mouse hover and construction button text.
  - fix: The ICBM launcher now has a proper build space geometry.
  - fix: The ICBM launcher is now revealed to the enemy like other super weapons.
  - fix: The ICBM launcher now uses the proper model on snow maps.
  - fix: The ICBM launcher no longer spawns stacked smoke columns below the building when damaged.
  - fix: The ICBM launcher now automatically repairs itself when idle like other USA buildings.
  - fix: The ICBM launcher can now be sabotaged by a GLA Saboteur (reset timer).
  - fix: The ICBM launcher now creates debris when destroyed while under construction.
  - fix: The ICBM description now displays the correct recharge time of 4 minutes.
  - fix: The ICBM now flies forwards instead of backwards and sideways.
  - fix: The ICBM no longer plays a duplicate siren sound at launch location when fired.
  - fix: The ICBM no longer has a slow door closing animation.
  - fix: The Cruise Missile launcher now uses a unique mouse hover, construction button and ability text.
  - fix: The Cruise Missile launcher now has a proper build space geometry.
  - fix: The Cruise Missile launcher no longer displays the broken GUI:SupW_CruiseMissile countdown timer text.
  - fix: The Cruise Missile launcher is now revealed to the enemy like other super weapons.
  - fix: The Cruise Missile launcher now uses the proper model on snow maps.
  - fix: The Cruise Missile launcher no longer spawns stacked smoke columns below the building when damaged.
  - fix: The Cruise Missile launcher now automatically repairs itself when idle like other USA buildings.
  - fix: The Cruise Missile launcher can now be shut down by a Microwave tank.
  - fix: The Cruise Missile launcher can now be sabotaged by a GLA Saboteur (reset timer).
  - fix: The Cruise Missile launcher now creates debris when destroyed while under construction.
  - fix: The Cruise Missile launcher no longer explodes with a mushroom cloud when destroyed.
  - fix: The Cruise Missile launcher now spawns 10 Rangers when destroyed.
  - fix: The Cruise Missile description now displays the correct recharge time of 2 minutes.
  - fix: The Cruise Missile now uses the cursor decal of the MOAB.
  - fix: The Cruise Missile now uses the proper model.
  - fix: The Cruise Missile now flies forwards instead of backwards and sideways.
  - fix: The Cruise Missile no longer plays a duplicate siren sound at launch location when fired.
  - fix: The Cruise Missile no longer has a slow door closing animation.
 
- [x] rotate x 270
- [x] rn ROCKPOST -> ROCKPOS (same as China)

xezon commented on Patch104pZH/GameFilesEdited/Data/INI/CommandSet.ini:
```diff
@@ -3915,7 +3915,9 @@ CommandSet SupW_AmericaDozerCommandSet
   7  = SupW_Command_ConstructAmericaPatriotBattery
   8  = SupW_Command_ConstructAmericaCommandCenter
   9  = SupW_Command_ConstructAmericaFireBase
+  10 = SupW_Command_ConstructAmericaNuclearMissileLauncher
```
Documentation missing.

commy2 replied:
Debug. Need a way to build it for testing.

xezon replied:
So Command Set will be reverted? The weapon can be tested on the [!TestAllFactionsBuildingsUnitsPromo](https://github.com/TheSuperHackers/GeneralsGamePatch/tree/main/Maps/Any/!TestAllFactionsBuildingsUnitsPromo) map.

commy2 replied:
Need to test the construction buttons as well. Both read ICBM. One button builds a "Cruise Missile", the other a "Nuclear Missile". Same model and portrait, but one shoots a missile with China Nuke (no radiation), the other a missile with MOAB explosion.

xezon commented on Patch104pZH/GameFilesEdited/Data/INI/Object/SuperWeaponGeneral.ini:
```diff
@@ -19169,6 +19178,15 @@ Object SupW_AmericaCruiseMissileLauncher
+  ; Patch104p @bugfix commy2 17/07/2023 Fix missile rotation when fired.
+  Draw = W3DModelDraw ModuleTag_MissileRotationFix
+    ConditionState = NONE
+      Model = ABNukeSilo_ROT
```
Where is ABNukeMissle referenced?

commy2 replied:
ABNukeMissle is an overwrite. The entire model has to be rotated. ABNukeSilo_ROT is a new model (just some bones).

xezon reviewed (commented):
yaml documentation missing

commy2 commented:
- fix geometry
- fixed Cruise Missile button image and cursor radius decal (cruise missile is not nuclear)
- fixed Cruise Missile description (similar to MOAB, 2 minute timer) Shortcut: C

todo
- damage states and death
- cruise missile and icbm construction button text and descriptions

commy2 commented:
Need to rotate the missile around z. Rocket direction doesn't match silo.

MTKing4 commented:
> - fixed Cruise Missile button image and cursor radius decal (cruise missile is not nuclear)

Curios what decal did you use?

commy2 commented:
MOAB, since it's the same explosion.

MTKing4 commented:
Oh ok

commy2 commented:
oof

  - fix: The ICBM launcher now uses a unique mouse hover and construction button text.
  - fix: The ICBM launcher now has a proper build space geometry.
  - fix: The ICBM launcher is now revealed to the enemy like other super weapons.
  - fix: The ICBM launcher no longer spawns stacked smoke columns below the building when damaged.
  - fix: The ICBM launcher now automatically repairs itself when idle like other USA buildings.
  - fix: The ICBM launcher can now be sabotaged by a GLA Saboteur (reset timer).
  - fix: The ICBM launcher now creates debris when destroyed while under construction.
  - fix: The ICBM description now displays the correct recharge time of 4 minutes.
  - fix: The ICBM now flies forwards instead of backwards and sideways.
  - fix: The ICBM no longer plays a duplicate siren sound at launch location when fired.
  - fix: The ICBM no longer has a slow door closing animation.
  - fix: The Cruise Missile launcher now uses a unique mouse hover, construction button and ability text.
  - fix: The Cruise Missile launcher now has a proper build space geometry.
  - fix: The Cruise Missile launcher no longer displays the broken GUI:SupW_CruiseMissile countdown timer text.
  - fix: The Cruise Missile launcher is now revealed to the enemy like other super weapons.
  - fix: The Cruise Missile launcher no longer spawns stacked smoke columns below the building when damaged.
  - fix: The Cruise Missile launcher now automatically repairs itself when idle like other USA buildings.
  - fix: The Cruise Missile launcher can now be shut down by an ECM tank.
  - fix: The Cruise Missile launcher can now be sabotaged by a GLA Saboteur (reset timer).
  - fix: The Cruise Missile launcher now creates debris when destroyed while under construction.
  - fix: The Cruise Missile launcher no longer explodes with a mushroom cloud when destroyed.
  - fix: The Cruise Missile launcher now spawns 10 Rangers when destroyed.
  - fix: The Cruise Missile description now displays the correct recharge time of 2 minutes.
  - fix: The Cruise Missile now uses the cursor decal of the MOAB.
  - fix: The Cruise Missile now uses the proper model.
  - fix: The Cruise Missile now flies forwards instead of backwards and sideways.
  - fix: The Cruise Missile no longer plays a duplicate siren sound at launch location when fired.
  - fix: The Cruise Missile no longer has a slow door closing animation.

--- 

I kept the Chinese build animation (dirt particles instead of USA sparks), because digging makes sense to me considering these are underground silos.

MTKing4 commented:
>   - fix: The Cruise Missile launcher can now be shut down by an ECM tank.

doesn't make sense, ECM Can't shut down buildings

commy2 commented:
Thanks for pointing that out. I meant the Microwave.

xezon commented:
Is this change ready for test?

commy2 commented:
Yeah, although the missile is rotated correctly yet. And I think I want to rename the ROCKPOST mempoint to ROCKPOS, to match China Nuclear Missile. I need to do this from home, because I don't have the magic Renx setup here that screws up the model the least.

commy2 commented:
PR complete now.

xezon commented:
Sweet. I will look later :-)

xezon commented:
Observations:

## ICBM tool tip

Says it is strong vs Buildings and Infantry. Why exactly that? Why not vehicles?

![shot_20230721_200322_4](https://github.com/TheSuperHackers/GeneralsGamePatch/assets/4720891/3d9bba17-9af7-4346-8ef6-ed75ed5e2d0c)

## Cruise tool tip

Cruise Missile says it will shock units, but it just makes a big explosion without shocks.

![shot_20230721_200317_3](https://github.com/TheSuperHackers/GeneralsGamePatch/assets/4720891/9cd4f3ab-76c2-4151-b982-02ec14ea835d)

## Cruise Prerequisite

The Cruise Missile requires a War Factory instead of a Strategy Center. Is this intentional?

![shot_20230721_204552_6](https://github.com/TheSuperHackers/GeneralsGamePatch/assets/4720891/318a9eba-ce95-4d4c-a47b-88d617dd9dd6)

## Audio loop bugs

The Audio loop of the ready ICBM and Cruise Missile is stuttering. Additionally the Audio loop does not restart when the camera scrolls away and back to the ICBM and Cruise Missile. The structure becomes muted. Both these issues do not happen with the Particle Cannon ready loop.

## Cruise Missile flight path

The flight path of the Cruise Missile is not good. It launches up and then flips angle in an instant. The Nuke Missile flight looks more gracious.

https://github.com/TheSuperHackers/GeneralsGamePatch/assets/4720891/32d74731-5729-49b3-9794-bc340ab30a41

## Cruise and ICBM Missile visuals

The Missile of both Superweapons changes visuals when launching.

https://github.com/TheSuperHackers/GeneralsGamePatch/assets/4720891/8362646e-aa0b-429e-91eb-1ca70a0fa33e

## Cruise and ICBM scaffold

There is orange smoke emitted from the construction scaffold. It is not clear why.

![shot_20230721_193741_2](https://github.com/TheSuperHackers/GeneralsGamePatch/assets/4720891/4e1d12ed-f6cf-40a3-b546-5944f3badc3e)

MTKing4 commented:
> The flight path of the Cruise Missile is not good. It launches up and then flips angle in an instant. The Nuke Missile flight looks more gracious.

It is probably because ICBM flies higher so you don't see that abrupt angle turn when it happenes

commy2 commented:
> Says it is strong vs Buildings and Infantry. Why exactly that? Why not vehicles?

Same as Nuclear Missile. Strong vs infantry presumably because of the radiation? Although ICBM has none.

> Cruise Missile says it will shock units, but it just makes a big explosion without shocks.

I copy pasted the MOAB for this. Should think about MOAB description anyway. I think "shocking" units is some dropped feature that only survives in the tooltip as misinfo.

> The Cruise Missile requires a War Factory instead of a Strategy Center. Is this intentional?

Didn't change that. That's how it works in 1.04.

> The Audio loop of the ready ICBM and Cruise Missile is stuttering. Additionally the Audio loop does not restart when the camera scrolls away and back to the ICBM and Cruise Missile. The structure becomes muted. Both these issues do not happen with the Particle Cannon ready loop.

The logic used is that of the Nuclear Missile. I would assume the same issues exist with it.

> The Audio loop of the ready ICBM and Cruise Missile is stuttering. Additionally the Audio loop does not restart when the camera scrolls away and back to the ICBM and Cruise Missile. The structure becomes muted. Both these issues do not happen with the Particle Cannon ready loop.

Values from base game. Maybe could be smoothed out, not sure.

> The Missile of both Superweapons changes visuals when launching.

Might be a material issue. I have no working setup that correctly imports and exports models. If you have, you need to rotate the missile around x by 270, and then recenter the model.

> There is orange smoke emitted from the construction scaffold. It is not clear why.

Those are the China scaffold particles. USA usually uses sparks. I would assume it uses the dirt effect, because most of this is copy pasted from the Nuclear Missile. I kept it, because dirt effect makes sense to me for digging a silo.

xezon commented:
I will look into Nuke Silo audio and MOAB tooltip issues.

xezon commented on Patch104pZH/GameFilesEdited/Data/INI/CommandSet.ini:
```diff
@@ -3931,8 +3931,9 @@ CommandSet SupW_AmericaDozerCommandSet
-  14 = Command_Stop
+  ;13 = Command_DisarmMinesAtPosition
+  ;14 = Command_Stop
+  14 = Command_DisarmMinesAtPosition
```
This does not look right.

xezon commented on Patch104pZH/GameFilesEdited/Data/INI/SpecialPower.ini:
```diff
@@ -818,11 +818,12 @@ End
   Enum                    = NUKE_SPECIAL_NEUTRON_MISSILE
   ReloadTime              = 300000   ; in milliseconds. min is 2x door/open close time!
-  InitiateSound           = AirRaidSiren  ; plays at source
+;  InitiateSound           = AirRaidSiren  ; plays at source
```
What does this do?

commy2 replied:
> no longer plays a duplicate siren sound at launch location when fired

It's the same as Nuclear Missile with this change.

xezon commented on Patch104pZH/GameFilesEdited/Data/generals.str:
```diff
@@ -92115,3 +92117,102 @@ RU: "Сильна против зданий\n\nВремя для готовно
+END
+
+CONTROLBAR:SupW_TooltipFireCruiseMissile
+US: "Strong vs. Buildings\nWill shock units in the blast radius\n\nCountdown Timer: 2:00"
```
You can update this text same as #2146

xezon commented on Patch104pZH/GameFilesEdited/Data/generals.str:
```diff
@@ -49631,18 +49631,20 @@ RU: "Силён против техники\n\nВремя для готовно
-PL: "Dobra przeciwko budynkom, piechocie\n\nCzas aktywacji: 6:00"
-RU: "Сильна против зданий и пехоты\n\nВремя для готовности: 6:00"
-AR: "ةﺄﺸﻤﻟا قﺮﻓو تﺄﺸﻨﻤﻟا ﺪﺿ لﺎﻌﻓ\n\n٠٠ : ٦ : ﻲﻟزﺎﻨﺘﻟا ﺪﻌﻟا ﺖﻗﺆﻣ"
+US: "Strong vs. Buildings, infantry\n\nCountdown Timer: 4:00"
```
I don't like this highlighting infantry over vehicles. It should say "Strong vs. Buildings, Vehicles". This text can be copied from `CONTROLBAR:TooltipFireTomahawkStorm`.

xezon replied:
It is also possible to write: "Strong vs. Everything" as seen in `CONTROLBAR:TooltipFireNukeMissile`.

xezon commented on Patch104pZH/GameFilesEdited/Data/INI/Object/SuperWeaponGeneral.ini:
```diff
@@ -20590,23 +19243,22 @@ Object SupW_AmericaCruiseMissileLauncher
-
-  Behavior        = FXListDie ModuleTag_21
+  Behavior             = CreateObjectDie ModuleTag_21
+    CreationList  = SupW_OCL_AmericanRangerDebris10
```
10 Rangers are a lot. The Particle Cannon just spawns 6. The Fusion Reactor 2. The Supply Drop Zone 3.

commy2 replied:
Oh, I thought it spawned 10. Guess that is only the CC. Same change should be made for Tomahawk Storm.

xezon reviewed (commented):
## Scaffolding particles

I think the orange China dust scaffolding effect needs to be replaced with the USA typical construction sparks. Can look at Supply Drop Zone for reference.

## Cruise Missile flight path

The flight path of the Cruise Missile needs another look. When not in this change, then in a follow up change.

## Damage particles

The damage particle effects are very large compared to the size of the structure. Perhaps they should use the damage particle effects of the Fusion Reactor or Supply Drop Zone.

![shot_20230725_192748_1](https://github.com/TheSuperHackers/GeneralsGamePatch/assets/4720891/0eb2c4a8-020a-43a5-8505-eebcdfb3b6f9)

![shot_20230725_192759_2](https://github.com/TheSuperHackers/GeneralsGamePatch/assets/4720891/6e3c2a6d-97c1-4965-a8cc-f3ffbd6360da)

xezon commented:
What do you mean by broken texture?

xezon commented on Patch104pZH/GameFilesEdited/Data/INI/CommandButton.ini:
```diff
@@ -7296,13 +7296,14 @@ CommandButton SupW_Command_ConstructAmericaNuclearMissileLauncher
   DescriptLabel           = CONTROLBAR:ToolTipAmericaICBMLauncher
 End
 
+; Patch104p @tweak commy2 20/07/2023 Add unique text and description.
```
Can add pull number here.

xezon commented on Patch104pZH/GameFilesEdited/Data/generals.str:
```diff
@@ -49631,18 +49631,20 @@ RU: "Силён против техники\n\nВремя для готовно
 AR: "تﺎﺑﺮﻌﻟا ﺪﺿ لﺎﻌﻓ\n\nﻖﺋﺎﻗد : ٣ : ﻲﻟزﺎﻨﺘﻟا ﺪﻌﻟا ﺖﻗﺆﻣ"
 END
 
+// Patch104p @fix commy2 20/07/2023 Fix timer tooltip (from 6:00).
```
Can add pull number here. No need to put (from 6:00) in parentheses.

xezon commented on Patch104pZH/GameFilesEdited/Data/generals.str:
```diff
@@ -92115,3 +92117,102 @@ RU: "Сильна против зданий\n\nВремя для готовно
 AR: "تﺄﺸﻨﻤﻟا ﺪﺿ لﺎﻌﻓ\n\n٠٠ : ٢ : ﻲﻟزﺎﻨﺘﻟا ﺪﻌﻟا ﺖﻗﺆﻣ"
 END
 
+// Patch104p @fix commy2 20/07/2023 Add missing strings for Cruise Missile and ICBM.
```
Can add pull number here.

xezon commented on Patch104pZH/GameFilesEdited/Data/INI/Object/WeaponObjects.ini:
```diff
@@ -5733,9 +5733,10 @@ Object CruiseMissile
   Behavior = NeutronMissileUpdate ModuleTag_03
     DistanceToTravelBeforeTurning = 200       ; goes straight up a long ways first
-    MaxTurnRate = 7200                        ; huge, since it turns off-camera
+    MaxTurnRate = 540 ; Patch104p @bugfix commy2 27/07/2023 Reduce MaxTurnRate to 540 from 7200. (#2122)
```
Generally you never need to specify new number in comment, because number is already on left side of comment.

xezon commented on Patch104pZH/GameFilesEdited/Data/INI/ObjectCreationList.ini:
```diff
@@ -9756,6 +9756,17 @@ ObjectCreationList AirF_OCL_AmericanRangerDebris05
   End
 End
 
+ObjectCreationList SupW_OCL_AmericanRangerDebris06
```
Can add a comment for adding.

xezon reviewed (approved):
Looks good. The tool tips still read bad, but is not a big deal. The Cruise Missile has no launch alarm and no launch, ready, build voice.

xezon commented on Patch104pZH/GameFilesEdited/Data/generals.str:
```diff
@@ -92115,3 +92117,102 @@ RU: "Сильна против зданий\n\nВремя для готовно
+END
+
+CONTROLBAR:SupW_TooltipFireCruiseMissile
+US: "Strong vs. Buildings\n\nCountdown Timer: 2:00"
```
Can fix capitalization same as #2150

commy2 commented:
Before me editing the model

![shot_20230730_135007_1](https://github.com/TheSuperHackers/GeneralsGamePatch/assets/6576312/887f60d8-c261-49e2-a552-3b0a4dbb601e)

(note visible bone bug)

After me editing the model

![shot_20230730_135104_1](https://github.com/TheSuperHackers/GeneralsGamePatch/assets/6576312/b2b04cce-2ad6-4672-824d-8c839c605fb3)

commy2 commented:
Missile changing appearance when launched is caused by it losing the shadow.

https://github.com/TheSuperHackers/GeneralsGamePatch/assets/6576312/9fa2e0dd-362f-4453-898f-541900455e98

Same happens for Nuclear Missile, Scud Storm and Tomahawk Storm.
